### ISSUE TITLE
Mod Inclusion Project: Tanks and Other Vehicles

### ISSUE THREAD

**Issue by Regularitee** [CleverRaven/Cataclysm-DDA#25502]
Why?

- In the 2016 Survey, it ranked no. 3 in the most used mods by the community. So its use is already extremely widespread.
- Tank drone overhauls require access to this mod's weaponry. Previously the 120mm cannon was imported manually from the mod and introduced into the base game. Now a second part, the 152mm ATGM also needs to be imported. At this point it may be more prudent to bring in the entire mod rather than one-at-a-time as needed, especially considering the needs of future robotic/military enemies that may yet be added to the game
- The high-caliber vehicle weapons have been balanced two weeks ago by me as described here (https://github.com/CleverRaven/Cataclysm-DDA/pull/25037), so balancing issues should all be already taken care of at this point.

Feel free to discuss the merits or demerits of this idea here.

**Comment by mlangsdorf** [comment]
See also #24730.
tl;dr: let's get treads working properly first.

**Comment by Regularitee** [comment]
Motherfucka mlangsdorf stole my thunder.

**Comment by mlangsdorf** [comment]
I would really rather not mainline treads in their current state. I had that option when I pulled parts of blazemod into mainline last month, and I deliberately didn't add the bandit bulldozer because I don't think treads work right.

There was at least some pushback against mainlining Tanks at all in 24730. Can we fix the issues with Tanks like the spawn rate of military vehicles while it remains a mod? It will still be available for the members of the player base that want it, and we won't be forcibly dragging unbalanced and too common tanks into the main game for people who don't want that.

**Comment by Regularitee** [comment]
>Could we please fix the issues with Tanks like the spawn rate of military vehicles while it remains a mod? 

You make it sound like it's something like this has a final answer, when in truth balance something that is fuzzy and requires rolling process. There will never be a perfect number we can set in stone; instead, it's a loop of feedback, tweaking, feedback and tweaking. Few mods have been feature-frozen since launch for good reason; developers aren't omniscient. 

>It will still be available for the members of the player base that want it, and we won't be forcibly dragging unbalanced and too common tanks into the main game for people who don't want that.

Well, as I mentioned, it was the 3rd most used mod int he 2016 survey. How many major PR changes had that exceptionally high amount of player utilization before they entered the mainstream? Just the opposite; of all the major PRs in the last year, it's probably the *least* likely of all to "forcibly drag" content onto players who "don't want that"

**Comment by kevingranade** [comment]
Bringing in the whole mod at once is a non-starter.
1. Its too big to review all at once.
2. I agree that treads should be fixed before we bring any tanks to mainline.
3. My understanding is that turrets are a trainwreck.
4. blob parts need careful review and discussion, I'm leaning toward leaving them out.
5. Solar arrays AFAIK completely throw any sense of proportionality for energy production out the window.

If you want to mainline stuff from blazemod, your best bet is to pick out the conventional stuff that doesnt require hacks and move them over a few pieces at a time.

**Comment by Regularitee** [comment]
4 & 5 are from "blazemod" and not "tanks and other vehicles", so not relevant to this topic.

3. I'm not sure trainwreck is the right word. From what I understand, turrets work and fire properly. There is one dummy part ("turret chassis") that currently does nothing, but otherwise turrets are fully functional... am I missing something for them to be considered a trainwreck?

**Comment by nexusmrsep** [comment]
But lets all agree here that we all want to eventualy have a tank mod in vanilla, even if it would be brought in part by part, with neccessary fixes along the way.

**Comment by Regularitee** [comment]
So what should be ported? Currently there are 11 items in the tanks mod:

- ammo_types 
- item_groups
- items
- ~~monster_drops~~
- ~~monsters~~
- parts
- recipes
- ~~snippets~~
- vehicle_groups
- vehicles
- ~~modinfo~~

Tankbots, as far as I know, are already incorporated in vanilla. So monsters and monster_drops are obsolete (correct me if I'm wrong here). Snippets are just harmless meme notes, and aren't really vital to the mod. Modinfo obviously won't be needed if it's no longer a mod. So that leaves:

- **ammo_types:** ammo obviously. recently had a balance overhaul, so safe to merge.
- **item_groups:** governs where/what vehicle ammo spawns. safe to merge.
- **items:** weapons. recently had balance overhaul, so safe to merge.
- **parts:** governs installing vehicle weapons. safe to merge.
- **recipes:** disassembling bots and assembling ammo. should be safe to merge. (concerns about whether the player should be able to make ammo should be its own issue, if anyone objects to that notion)

The only two that may cause concern (because they include treads, and therefore have blazemod dependencies) are:
- vehicle_groups
- vehicles

We can either remove all treaded vehicles from the list, or else manually edit all tracked vehicles to remove tracks, or maybe not have vehicles at all from the mod.

**Comment by mlangsdorf** [comment]
If you want to add a bunch of wrecked tanks as spawn, similar to the existing helicopter wreckage, go write that PR.  I don't see any value in merging Tanks by making the vehicles into wrecks.

I also quibble with the idea that the weapons, parts, and ammo_types had a balance overhaul and are thus safe to merge. They were overhauled with an eye toward realism and are less unbalanced than they used to be (thank you for doing that), but I certainly didn't do an exhaustive comparison of them against existing weapons to decide if the numbers were sensible or not.

My counter-proposal is:
1. Using the existing vehicle prototypes as placeholders, determine how many vehicles with each prototype we want to spawn and get the vehicle groups correct in the current mod.
2. Go back through the weapons and ammos and review them closely.
3. Revise the vehicle prototypes as we think is appropriate, such as merging the atomic tank and light tank, adding support vehicles like Combat Engineering Vehicles, etc, again making the changes in the mod.
4. When the mod is in better shape and has better support in code, merge the bits that need to merged.

I just don't get the rush to merge Tanks.

**Comment by Regularitee** [comment]
To be honest, I think we need to release these parts to test them properly. This game does not have vehicle-on-vehicle combat, so there are no "simple" results to test for like the thickness of armor penetration or spalling.

Instead, we have extremely fuzzy results based on damage to hordes or zombies. Is killing X zombies, maiming Y, and missing Z, overpowered or underpowered compared to killing A zombies, maiming B zombies, and missing C zombies? And how does that balance against the fact it's a vehicle turret and should be more power than conventional weapons? And further more, how does that balance against how scarce ammo is?

I'm worried an attempt at closed comprehensive testing could setback potential development of other game aspects by hours, days, even weeks.

**Comment by kevingranade** [comment]
> But lets all agree here that we all want to eventualy have a tank mod in vanilla, even if it would be brought in part by part, with neccessary fixes along the way.

Absolutely.

> ammo_types: ammo obviously. recently had a balance overhaul, so safe to merge.

I agree with mark, this was rebalanced to make it not ludicrous, not to the point of being mainline-ready. It could need anything between careful review and total overhaul before it's ready.

> item_groups: governs where/what vehicle ammo spawns. safe to merge.

If you mean it can be merged after adjusting spawn r as tes, probably, if you mean with current spawn rates almost certainly not.

> items: weapons. recently had balance overhaul, so safe to merge.

Same as ammo types, has not been reviewed for mainline inclusion at all.

> parts: governs installing vehicle weapons. safe to merge.

I anticipate a lot of issues with expanding vehicle weapons, it's already very expansive, and pushing it even further is going to meet with some resistance.

> recipes: disassembling bots and assembling ammo. should be safe to merge. (concerns about whether the player should be able to make ammo should be its own issue, if anyone objects to that notion)

Which robots? Regardless, yields need to be reviewed.  I'm not familiar with which ammo recipes these are, if it's more complicated than packing propellant into a shell, then it might not be acceptable.

> I'm worried an attempt at closed comprehensive testing could setback potential development of other game aspects by days, even weeks.

Nope, it'd just stall tank mod inclusion, which is fine, we dont have a deadline.

> Especially when open testing would result in much more robust and thorough results

This makes no sense, if theres interest in testing, it can happen *now*, if theres no interest in testing it wont happen regardless of being mainlined.
Plus you're missing the point, it doesnt need, "see if it visibly breaks something testing", it needs "review for fit with the base game review".

**Comment by Regularitee** [comment]
>If you mean it can be merged after adjusting spawn rates, probably, if you mean with current spawn rates almost certainly not.

What do you mean? Those spawn schedules are all for the cargo of tracked military vehicles. Vehicles which are not being mainlined because they have tracks. Ergo, nothing will be spawned -- except one thing, a single textbook. Is there something wrong with the textbook's spawn rates?

If you did mean the ammo, from what I've experienced, these vehicles generally only contain a single "clip" of ammo for a given turret, never more. Could you be more specific about what's "almost certainly not" acceptable about that? I mean, is it too high? Too low? It's kind of hard to address an issue that isn't made clear. 

>Nope, it'd just stall tank mod inclusion, which is fine, we don't have a deadline.
The [opportunity cost](https://en.wikipedia.org/wiki/Opportunity_cost) of spending of testing something as vague as balance. Debugging is one thing, but balance means attempting to account for the limitless possible scenarios, tactics, and ideas that an entire community would come up with, which is precisely why multiplayer betas exist in even the largest budget games.
